### PR TITLE
Add bindings for GstToc/TocSetter

### PR DESCRIPTION
**Pull request by fengalin**
See #10

### CODE REVIEW

File: gstreamer/src/event.rs [hunk 1]
```diff
@@ -468,8 +468,18 @@ impl<'a> StreamGroupDone<'a> {
...
-// TODO
 pub struct Toc<'a>(&'a EventRef);
+impl<'a> Toc<'a> {
+    pub fn get_toc(&self) -> (&'a ::TocRef, bool) {
+        unsafe {
+            let mut toc = ptr::null_mut();
+            let mut updated = mem::uninitialized();
+
+            ffi::gst_event_parse_toc(self.0.as_mut_ptr(), &mut toc, &mut updated);
+            (::TocRef::from_ptr(toc), updated != 0)
```
Comment (sdroege) [inline gstreamer/src/event.rs]: from_glib(updated)

File: gstreamer/src/event.rs [hunk 2]
```diff
@@ -932,26 +942,25 @@ impl EosBuilder {
...
     seqnum: Option<u32>,
     running_time_offset: Option<i64>,
-    #[allow(unused)]
-    toc: (),
+    toc: Option<::Toc>,
     updated: bool,
 }
 impl TocBuilder {
-    pub fn new(toc: (), updated: bool) -> Self {
+    pub fn new(toc: ::Toc, updated: bool) -> Self {
```
Comment (sdroege) [inline gstreamer/src/event.rs]: Could also take a &Toc as it's to_glib_none() anyway, but not too important

File: gstreamer/src/lib.rs [hunk 3]
```diff
@@ -100,6 +100,9 @@ pub use value::*;
 mod segment;
 pub use segment::*;
 
+pub mod toc;
+pub use toc::*;
```
Comment (sdroege) [inline gstreamer/src/lib.rs]: Instead of * only Toc and TocRef IMHO

File: gstreamer/src/message.rs [hunk 4]
```diff
@@ -878,7 +878,14 @@ impl<'a> Progress<'a> {
 
 pub struct Toc<'a>(&'a MessageRef);
 impl<'a> Toc<'a> {
-    // TODO get_toc()
+    pub fn get_toc(&self) -> (::Toc, bool) {
+        unsafe {
+            let mut toc = ptr::null_mut();
+            let mut updated = mem::uninitialized();
+            ffi::gst_message_parse_toc(self.0.as_mut_ptr(), &mut toc, &mut updated);
+            (from_glib_full(toc), updated != 0)
```
Comment (sdroege) [inline gstreamer/src/message.rs]: from_glib(updated)

File: gstreamer/src/toc.rs [hunk 5]
```diff
@@ -0,0 +1,350 @@
+use std::ffi::CStr;
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: Copyright header blurb missing

File: gstreamer/src/lib.rs (hunk 3 above)
Comment (sdroege) [inline gstreamer/src/lib.rs]: And TocEntry / TocEntryRef

File: gstreamer/src/toc.rs [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+        unsafe {
+            let toc_entry = ffi::gst_toc_find_entry(
+                self.as_ptr(),
+                uid.to_glib_none().0,
+            );
+            if toc_entry.is_null() {
+                return None;
+            }
+
+            Some(TocEntry::from_glib_borrow(
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: This is unsafe. Either return a &TocEntryRef or from_glib_none()

from_glib_borrow() return values must only be used in this very scope (and not be passed outside) and only references be passed to functions called from this scope

File: gstreamer/src/toc.rs [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+    pub fn get_tags(&self) -> Option<TagList> {
+        unsafe {
+            let tags = ffi::gst_toc_get_tags(self.as_ptr());
+            if tags.is_null() {
+                return None;
+            }
+
+            Some(TagList::from_glib_borrow(tags as *const ffi::GstTagList))
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: Same here

File: gstreamer/src/toc.rs [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+            from_glib_full(ffi::gst_mini_object_copy(self.as_ptr() as *const _) as
+                *mut _)
+        }
+    }
+}
+
+unsafe impl Sync for TocRef {}
+unsafe impl Send for TocRef {}
+
+
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: Double newline

File: gstreamer/src/toc.rs [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+    pub fn get_parent(&self) -> Option<TocEntry> {
+        unsafe {
+            let parent = ffi::gst_toc_entry_get_parent(self.as_mut_ptr());
+            if parent.is_null() {
+                return None;
+            }
+
+            Some(TocEntry::from_glib_borrow(
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: No borrow

File: gstreamer/src/toc.rs [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+    pub fn get_start_stop_times(&self) -> Option<(i64, i64)> {
+        unsafe {
+            let mut start = mem::uninitialized();
+            let mut stop = mem::uninitialized();
+
+            if ffi::gst_toc_entry_get_start_stop_times(
+                self.as_ptr(),
+                &mut start,
+                &mut stop
+            ) != 0
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: from_glib() for the bool conversion

File: gstreamer/src/toc.rs [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+    pub fn get_tags(&self) -> Option<TagList> {
+        unsafe {
+            let tags = ffi::gst_toc_entry_get_tags(self.as_ptr());
+            if tags.is_null() {
+                return None;
+            }
+
+            Some(TagList::from_glib_borrow(tags as *const ffi::GstTagList))
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: No borrow

**Comment by sdroege** [comment]
Generally looks good, just a few small things :) Thanks! What file are you testing that with currently?

**Comment by fengalin** [comment]
Thank you for your review and comments! I think I addressed all of them.
I used a couple mkv files and flac songs from an album that comes with an m3u playlist.

**Comment by fengalin** [comment]
I'm [building an application](https://github.com/fengalin/media-toc) to edit or generate toc from media files.
I gave a try to ffmpeg Rust binding and now completed the same with gstreamer.

**Comment by sdroege** [comment]
I shortly looked at your application, and https://github.com/fengalin/media-toc/blob/gstreamer/src/media/audio_buffer.rs looks like it could make use of GstAudioInfo (which I'm currently working on), and https://github.com/fengalin/media-toc/blob/gstreamer/src/media/media_info.rs of GstDiscoverer (which is also somewhere on my list but should be relatively easy to do)

I'll review your changes in a bit, thanks for the fast update :)

**Comment by fengalin** [comment]
Yes, I saw your work on `GstAudioInfo`. Will use it for sure.
Thanks for pointing me to `GstDiscover`.

File: gstreamer/src/event.rs [hunk 12]
```diff
@@ -942,25 +942,24 @@ impl EosBuilder {
...
-            toc: Some(toc),
+            toc: toc,
             updated: updated,
         }
     }
 
     event_builder_generic_impl!(|s: &mut Self| {
-        let toc = s.toc.take();
-        ffi::gst_event_new_toc(toc.to_glib_none().0, s.updated.to_glib())
+        ffi::gst_event_new_toc(s.toc.to_glib_none().0, s.updated.to_glib())
```
Comment (sdroege) [inline gstreamer/src/event.rs]: &mut Self not needed here, can be just &Self now

File: gstreamer/src/toc.rs [hunk 13]
```diff
@@ -214,7 +221,7 @@ impl TocEntryRef {
                 return None;
             }
 
-            Some(TagList::from_glib_borrow(tags as *const ffi::GstTagList))
+            Some(TagList::from_glib_none(tags as *const ffi::GstTagList))
```
Comment (sdroege) [inline gstreamer/src/toc.rs]: FWIW, this is exactly why the Ref types for the miniobjects exist and you can borrow them. But this seems not needed here and more complicated than anything else.

**Comment by sdroege** [comment]
Looks good now except for the additional &mut

File: gstreamer/src/event.rs (hunk 12 above)
Comment (fengalin) [inline gstreamer/src/event.rs]: Yes! Sorry about that.

Comment (sdroege) [inline gstreamer/src/event.rs]: No problem, easy to forget these little things :)

File: gstreamer/src/toc.rs (hunk 13 above)
Comment (fengalin) [inline gstreamer/src/toc.rs]: Do you mean something like this?
``` Rust
    pub fn get_tags(&self) -> Option<&TagListRef> {
        unsafe {
            let tags = ffi::gst_toc_get_tags(self.as_ptr());
            if tags.is_null() {
                return None;
            }

            Some(TagListRef::from_ptr(tags as *const ffi::GstTagList))
        }
    }
```
It seems weird to return `&TagListRef`.

Comment (sdroege) [inline gstreamer/src/toc.rs]: Yes, that's why I said it seems unneeded :) It will work, but just keep it as-is for now.

**Comment by fengalin** [comment]
Just commited the `&mut` part. I can still work on the other issue in order to get it right.

File: gstreamer/src/toc.rs (hunk 13 above)
Comment (fengalin) [inline gstreamer/src/toc.rs]: All right!

**Comment by fengalin** [comment]
Add the last one if you deem it to be useful :)

**Comment by sdroege** [comment]
Thanks, merged :)

**Comment by sdroege** [comment]
AudioInfo bindings are there now, let me know if that makes your life a bit easier or if something else is missing :)

**Comment by fengalin** [comment]
Thank you! That will be much better than my "naive" implementation. I will also use `AudioFormatInfo` `unpack` function in order to replace my own sample conversions to `f64`.

**Comment by sdroege** [comment]
Unpacking will go to S32 for integer input formats though. You either want GstAudioConverter, or you could just let an audioconvert element at the end of your pipeline convert to f64. For that you could here (https://github.com/fengalin/media-toc/blob/be11572f21134efe796c87804ab1d2ce11ba11c4/src/media/context.rs#L135) have a pipeline with a tee, one branch going to an autoaudiosink like now, another going to an audioconvert and appsink for collecting the frames and f64 conversion.

**Comment by fengalin** [comment]
Yes, I thought about a tee too. Integer samples would be good enough though. I only need a predictable range and 0 offset in order to [draw the waveform](https://github.com/fengalin/media-toc/blob/6aa9ad106603a0990a87b38a3921408998bb515b/src/ui/audio_controller.rs#L76).

I though about it and now my concern is to get a good balance between the time spent in GStreamer's thread in the pad probe callback and the time spent in the drawing area `draw` event. Given that the `draw` event will be executed more often (every redraw), I tend to think I need to reduce the computation needed there. That's the reason why I [uninterleaved the channels and normalized the samples](https://github.com/fengalin/media-toc/blob/6aa9ad106603a0990a87b38a3921408998bb515b/src/media/audio_buffer.rs#L161) - well I got to the point that I didn't need to interleave channels actually since seeking in an array is in o(1) -. On the other end, I don't want to hang the GStreamer thread too long and cause lags. I suspect that with current hardware, that should not happen anyway. This is interesting, I'd like to find a good operating point, but I guess it is really system dependent.

**Comment by sdroege** [comment]
That's also why I suggested the tee. Basically what you would have is:
```
          --- queue --- audioconvert --- appsink
---- tee /
         \
          --- queue --- audioconvert --- audioresample --- autoaudiosink
```

The thread where you get the audio buffers in the appsink is a different one than the one that goes to the audiosink (the queue is a thread barrier), so you can spend more time doing things there without interrupting playback (as long as everything still runs in real-time on average). You should be able to completely prepare the samples there to whatever you need for the rendering in the UI thread later, so that the UI thread only has to render and not do any calculations anymore that are unrelated to the visualization.

You would then get e.g. F64 in the appsink, could deinterleave there as needed, normalize samples, etc and just have the amplitudes of the waveform you want to draw as a result there.

**Comment by fengalin** [comment]
Great! I'll do that. Many thanks for the wise advices.

**Comment by sdroege** [comment]
https://github.com/sdroege/gstreamer-rs/commit/b655c838b285b4dea8dff4f36a2c340b092b8b66 is probably also useful for you, to simplify things even further

**Comment by fengalin** [comment]
Yes indeed! Thanks again!

**Comment by sdroege** [comment]
API might still minimally change, but that should be no problem to use it :)

**Comment by fengalin** [comment]
No problem! I'll give you a feedback soon.

**Comment by sdroege** [comment]
Great thanks, let me know if something could be nicer

**Pull request closed without merging**